### ISSUE TITLE
Install onto vanilla termux?

### ISSUE THREAD

**Issue by KoviRobi** [t184256/nix-on-droid#257]
Hi,

Is it possible to install onto vanilla termux? The main reason I would like to do so is for using solarized colorschemes (if I do Nix-on-droid -> [hold] -> More... -> Style, it says "The Termux:Styling Plugin App is not installed", despite it being installed (via F-Droid).

I have Nix v0.118.0_v0.3.1.

Probably relevant:
- https://github.com/t184256/nix-on-droid/issues/4 
- https://github.com/t184256/nix-on-droid/issues/56

I think there might be other upsides to installing onto termux:
- Updates from termux app
- It comes with `am` installed, which would help
  - https://github.com/t184256/nix-on-droid/issues/221 
  -  https://github.com/t184256/nix-on-droid/issues/105

I think the install is done by

https://github.com/t184256/nix-on-droid-app/blob/f55587fc9eed7d8a0515f05f0b40a8b0c72c7c66/app/src/main/java/com/termux/app/TermuxInstaller.java#L149-L285

which as I understand unzips, creates symlinks, chmods executables. But not sure what it does after to start the whole thing

**Comment by KoviRobi** [comment]
Ok I am getting there:
- extract bootstrap and `cd` into the extracted folder (`unzip nix-on-droid-bootstrap-aarch64.zip -d nix-on-droid-bootstrap-aarch64` and `cd nix-on-droid-bootstrap-aarch64`)
- fixup `bin/login` and `usr/lib/login-inner` by replacing `com.termux.nix` -> `com.termux`
- copy files over `cp -r -v --no-clobber */ /data/data/com.termux/files/usr/`

**Comment by t184256** [comment]
1. Yes: https://github.com/t184256/nix-in-termux
2. For colorschemes alone, you don't have to, just copy one over from Termux's `.termux/colors.properties`.

**Comment by KoviRobi** [comment]
Thanks! I did spot that, but I had the `invalid ELF header` with that, and I saw it is archived so I'd try this, but in the end I had it here too. I did manage to solve that though:

https://github.com/KoviRobi/nix-on-droid/commit/95c17519f830ff2dba58ec5aac985065386755d7

So now the steps are:
```console
[pc:~/nix-on-droid]$ NIX_ON_DROID_CHANNEL_URL=https://github.com/KoviRobi/nix-on-droid/archive/refs/heads/install-on-vanilla-termux.tar.gz \
                     NIX_ON_DROID_FLAKE_URL=github:KoviRobi/nix-on-droid/install-on-vanilla-termux \
                     nix build --impure .#bootstrapZip
[android:~]$ scp pc:nix-on-droid/result/bootstrap-aarch64.tar.gz .
[android:~]$ mkdir ../nix/
[android:~]$ tar xf ~/bootstrap-aarch64.tar.gz -C ../nix/
[android:~]$ chmod +w -R ../nix/
[android:~]$ ../nix/bin/login
```

which works for me :) Only one thing remaining is after install it downgrades nix to 2.11.1 which has this issue https://github.com/t184256/nix-on-droid/issues/213

**Comment by Gerschtli** [comment]
@KoviRobi That's pretty neat. Is the termux integration with its addons and the `am` script working fine?

**Comment by KoviRobi** [comment]
So there's some gotcha's I haven't figured out:

The `-b /data/data/com.termux/files/nix/bin:/bin` part of proot (from `nix/bin/login`) makes termux programs break, e.g.
```console
~ $ /data/data/com.termux/files/usr/bin/bash
-bash: /data/data/com.termux/files/usr/bin/bash: cannot execute: required file not found
```

which in turn breaks stuff because `$SHELL` in termux by default is set to said bash, e.g.
```console
~ $ ../nix/bin/login
Welcome to Nix-on-Droid!
If nothing works, open an issue at https://github.com/t184256/nix-on-droid/issues or try the rescue shell.
~ $ nix shell nixpkgs#iproute2
error: unable to execute '/data/data/com.termux/files/usr/bin/bash': No such file or directory
~ $ unset SHELL
~ $ nix shell nixpkgs#iproute2
bash-5.2$ ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
[...]
```

If instead I remove/comment out (taking care of `bash` not liking commenting out lines with line-continuaton `\`)
```console
~ $ ../nix/bin/login
proot warning: can't sanitize binding "/linkerconfig": Permission denied
Welcome to Nix-on-Droid!
If nothing works, open an issue at https://github.com/t184256/nix-on-droid/issues or try the rescue shell.
~ $ $SHELL
# works
~ $ /data/data/com.termux/files/usr/bin/bash
# works
```

But `am` doesn't yet work
```console
~ $ ../nix/bin/login
Welcome to Nix-on-Droid!
If nothing works, open an issue at https://github.com/t184256/nix-on-droid/issues or try the rescue shell.
~ $ am
-bash: /data/data/com.termux/files/usr/bin/am: cannot execute: required file not found
~ $ strace am
[...]
openat(AT_FDCWD, "/system/lib64/libnativeloader.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/odm/lib64/libnativeloader.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/vendor/lib64/libnativeloader.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
[...]
write(2, "library \"libnativeloader.so\" not"..., 65library "libnativeloader.so" not found: needed by main executable) = 65
[...]
exit_group(1)                           = ?
+++ exited with 1 +++
```
compared with vanilla termux
```console
~ $ strace am
openat(AT_FDCWD, "/system/lib64/libnativeloader.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/system_ext/lib64/libnativeloader.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/apex/com.android.art/lib64/libnativeloader.so", O_RDONLY|O_CLOEXEC) = 9
readlinkat(AT_FDCWD, "/proc/self/fd/9", "/apex/com.android.art/lib64/libn"..., 4096) = 46
```

I believe this is due to
```console
newfstatat(AT_FDCWD, "/linkerconfig/ld.config.txt", {st_mode=S_IFREG|0644, st_size=71632, ...}, 0) = 0
openat(AT_FDCWD, "/linkerconfig/ld.config.txt", O_RDONLY|O_NOFOLLOW|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=71632, ...}) = 0
mmap(NULL, 73728, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7c05926000
prctl(PR_SET_VMA, PR_SET_VMA_ANON_NAME, 0x7c05926000, 73728, "bionic_alloc_lob") = 0
read(3, "dir.system = /system/bin/\ndir.sy"..., 1024) = 1024
[...]
openat(AT_FDCWD, "/dev/__properties__/u:object_r:vndk_prop:s0", O_RDONLY|O_NOFOLLOW|O_CLOEXEC) = 3
```

but doing `-b /linkerconfig:/linkerconfig` doesn't seem to work. More experimentation needed

**Comment by KoviRobi** [comment]
But the termux addons seem to work (tried termux:style, and termux:float anyway)

**Comment by KoviRobi** [comment]
I will probably pursue this further still, it doesn't feel like it's far now :)

**Comment by KoviRobi** [comment]
Success! I had to remove the following lines from the proot script
```
 -b /data/data/com.termux/files/nix/bin:/bin \
 -b /data/data/com.termux/files/nix/etc:/etc \
```
and instead of using the `proot-static`, I used the `proot` from termux.

And I could do the following:
```
~ $ ../nix/bin/login
Welcome to Nix-on-Droid!
If nothing works, open an issue at https://github.com/t184256/nix-on-droid/issues or try the rescue shell.
~ $ am broadcast --user 0 \
                 --es com.termux.app.reload_style storage \
                 -a com.termux.app.reload_style com.termux
```
and it asked for permissions to access my storage :)

**Comment by t184256** [comment]
> and instead of using the proot-static, I used the proot from termux.

Huh. Never thought that's what I've been missing. Our proot is compiled from their sources.

**Comment by t184256** [comment]
Tried not mounting over `/bin` and `/etc`, that wasn't enough indeed. Will try with newest proot-termux patches next time.

**Comment by KoviRobi** [comment]
I should also say not mounting /etc means nix has problems due to lack of resolv.conf (I think) so I need to look into that too

**Comment by t184256** [comment]
Yeah, I realize that, I just hope it'll only need some linker configuration or something equally trivial.

**Comment by t184256** [comment]
> doing -b /linkerconfig:/linkerconfig doesn't seem to work

https://github.com/termux/proot/commit/0fdcda1d302dd05c175f250f2dd06c90c4e255cb is the right `-b /linkerconfig`

**Comment by t184256** [comment]
`~ $ am broadcast`

which `am` was it, `/system/bin/am`, `/bin/am`, or termux's special `am`?

**Comment by KoviRobi** [comment]
```
~ $ nix run --no-net nixpkgs#bashInteractive
bash-5.2$ am broadcast --user 0 --es com.termux.app.reload_style storage -a com.termux.app.reload_style com.termux
Broadcasting: Intent { act=com.termux.app.reload_style pkg=com.termux (has extras) }
Broadcast completed: result=0
bash-5.2$ command -v am
/data/data/com.termux/files/usr/bin/am
```

**Comment by myguidingstar** [comment]
@t184256 @KoviRobi 
Has anyone tried termux's very own `proot-distro`? It seems to allow adding custom distro by providing a tarball https://github.com/termux/proot-distro#adding-distribution

**Comment by tcely** [comment]
> @t184256 @KoviRobi Has anyone tried termux's very own `proot-distro`? It seems to allow adding custom distro by providing a tarball https://github.com/termux/proot-distro#adding-distribution

I'd like to see `nix-on-droid` have a configuration that I could just add to Termux.

https://github.com/termux/proot-distro/issues/215#issuecomment-1397017046

It looks like they went wrong by trying to support multiple users (with the daemon) but a single user system would make more sense for Termux and Android.

**Comment by SeerLite** [comment]
@KoviRobi what ended up missing? Would following your instructions in Termux give a working Nix setup?

**Comment by KoviRobi** [comment]
I can't remember exactly, it was a weekend project kind of thing which ran out of steam because even with using a monitor for my phone, and plugging in a keyboard, my phone wasn't ready to just replace my laptop (when travelling).

But also my motivation for this as opposed to the nix app was terminal styling, which turned out you can manually edit the `.termux/colors.properties`.

You are welcome to try and maybe you can fix the last couple of issue if you want

**Comment by KoviRobi** [comment]
Oh btw, I have a setup[^1] again when I use my phone as a laptop (and do more commuting), so I came back to this, I rebased, and everything works normally as I expected?

I rebased from my comment https://github.com/nix-community/nix-on-droid/issues/257#issuecomment-1465341185 to get https://github.com/KoviRobi/nix-on-droid/commit/101d81c960d3b0a31982673365f5282d2638b16a

Then to use, I did the same steps as from the comment
```console
[pc:~/nix-on-droid]$ NIX_ON_DROID_CHANNEL_URL=https://github.com/KoviRobi/nix-on-droid/archive/refs/heads/install-on-vanilla-termux.tar.gz \
                     NIX_ON_DROID_FLAKE_URL=github:KoviRobi/nix-on-droid/install-on-vanilla-termux \
                     nix build --impure .#bootstrapZip
[android:~]$ scp pc:nix-on-droid/result/bootstrap-aarch64.tar.gz .
[android:~]$ mkdir ../nix/
[android:~]$ tar xf ~/bootstrap-aarch64.tar.gz -C ../nix/
[android:~]$ chmod +w -R ../nix/
[android:~]$ ../nix/bin/login
```

I should check if some of the other commits are even necessary on my fork now.

It is noticeably slower (about 4x slower for some OCaml task, definitely not a scientific measurement), I suspect that's due to `proot`? So nix-on-droid would have the same issue too.

I should compare that with nix-on-droid too.

[^1]: Setup consists of phone, some AR glasses (XReal Air 2 Pro), bluetooth keyboard, bluetooth trackball, and I'm able to work on a 2h total daily commute reasonably comfortably.

**Comment by izelnakri** [comment]
@KoviRobi **FANTASTIC WORK!** After spending DAYS making nix work on Termux, only your last comment snippet and github fork/branch worked for me. I even failed at making this work on my own with my own fork of nix-on-droid with changes!

@t184256 I've been using happily using nix-on-droid for over a year now, however yesterday I tried to dig deeper and tried to run Nix-On-Droid app binaries on **device boot** AND be able to utilize Termux RUN_COMMAND Android Intent(RPC framework) in my other android apps, both of which failed! 

**WE ABSOLUTELY need this feature directly supported by this project**, my recent experiences(after one year) made me realize having a Termux fork might be a mistake now(experiment/legacy reasons) and we should instead distribute it via `$ pkg install nix` or `$ pkg install nix-on-droid`.

Why we need to transform this project or **at least directly support a binary release of the bootstrap script for Termux**:

a - Nix-on-droid app uses a fork of Termux which backtracks new features, changes & fixes to current Termux, if it ever gets these changes.
b - Nix-on-droid app doesn't support Termux addons which are essential if you want to run scripts/binaries in your shell environment *on device boot* (Termux:Boot) or service management(like previously discussed supervisord, runit or systemd: Termux:Services)
c - Nix-on-droid app doesn't support Termux RUN_COMMAND Intent interface which is required if you want to run a binary installed on users system from users another android app via Android Intent RPC framework.
d - Nix-on-droid repo targets `com.termux.nix` not `com.termux`, installing `nix-on-droid` on Termux is NOT trivial, currently whereas it could.
e - Nix-on-droid gets outdated documentation due to Termux fork legacy and simply not updating some docs to Nix-on-Droid.
f - Updating `nix` binary/channel is an error-prone operation.

Let me know @t184256 if I can also help in any way, if contribution is needed, I think we should rather distribute this amazing solution that you maintain as a Termux package going forward.

Android is quickly becoming the IOT platform Operating System of the future. I recently bought a VR headset Meta Quest 3 with 512GB storage to use it as my nix development machine, paying only 24 EUR a month. My smart watch and foldable phone runs on android as well. Although `proot` slows things down significantly, `nix-on-droid` is well versed to become an essential tool to convert android devices into development machines.

Going forward, I'll also try to set up systemd, docker with LXC, maybe then I can have everything I need for my android devices. @t184256  Thank you for working on this project, I hope we can maintain & improve it further!

**Comment by t184256** [comment]
You're free to either send PRs to the -app repo that'd implement the missing stuff (that I don't understand because I'm not an android dev) and/or to start a competing thing that lives inside termux. That's actually how I've started. But I won't develop that one, I barely have the energy for this project.

Also, I sincerely hope the long-term future is neither, it's real NixOS virtual machines on Android. But I don't own a single device that has hardware acceleration on Android yet.

**Comment by izelnakri** [comment]
I agree with you on NixOS with android runtime like waydroid, but let's be realistic for now, hardware OEMs like Samsung, Meta and even Google won't like that ;) I hope you reconsider my request for collaboration! ;)
